File: Python/108-convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        def BSTBuild(start: int, end: int) -> TreeNode:
            if start > end:
                return None
            mid = (start + end) // 2
            root = TreeNode(nums[mid])
            #递归构造
            root.left = BSTBuild(start, mid - 1)
            root.right = BSTBuild(mid + 1, end)
            return root
        if nums == []:
            return None
        return BSTBuild(0, len(nums) - 1)

File: Go/108-convert-sorted-array-to-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func BSTBuild(nums []int, start, end int) *TreeNode {
    if start > end {
        return nil
    }
    mid := (start + end) / 2
    root := &TreeNode{
        Val: nums[mid],
        Left: nil,
        Right: nil,
    }
    root.Left = BSTBuild(nums, start, mid - 1)
    root.Right = BSTBuild(nums, mid + 1, end)
    return root
}

func sortedArrayToBST(nums []int) *TreeNode {
    return BSTBuild(nums, 0, len(nums) - 1)
}